From 85601c5fa10002e6b8f56adcec14f634911f0c00 Mon Sep 17 00:00:00 2001
From: Tang Yun ping <typ@rock-chips.com>
Date: Thu, 24 Aug 2017 11:52:50 +0800
Subject: [PATCH] PM / devfreq: rockchip-dmc: using drm_modeset mutex when get
 lcdc type

Because ddr freq scanning may get clk mutex than get drm general
mutex to get lcdc_type. But when drm status change, drm may get drm
general mutex first and than get clk mutex. This may cause deadlock.
So when get lcdc type we change to use drm_modeset mutex to avoid
deadlock.

Change-Id: I64d4c236b682d7bd5bdc95264ff2711ebcec2cad
Signed-off-by: Tang Yun ping <typ@rock-chips.com>
---
 drivers/devfreq/rockchip_dmc.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/devfreq/rockchip_dmc.c b/drivers/devfreq/rockchip_dmc.c
index 13a0a5bf2cf2..f450ae3031a3 100644
--- a/drivers/devfreq/rockchip_dmc.c
+++ b/drivers/devfreq/rockchip_dmc.c
@@ -16,6 +16,7 @@
 #include <dt-bindings/display/rk_fb.h>
 #include <dt-bindings/soc/rockchip-system-status.h>
 #include <drm/drmP.h>
+#include <drm/drm_modeset_lock.h>
 #include <linux/arm-smccc.h>
 #include <linux/clk.h>
 #include <linux/cpu.h>
@@ -597,14 +598,14 @@ static int rk_drm_get_lcdc_type(void)
 	if (drm) {
 		struct drm_connector *conn;
 
-		mutex_lock(&drm->mode_config.mutex);
+		drm_modeset_lock(&drm->mode_config.connection_mutex, NULL);
 		drm_for_each_connector(conn, drm) {
 			if (conn->encoder) {
 				lcdc_type = conn->connector_type;
 				break;
 			}
 		}
-		mutex_unlock(&drm->mode_config.mutex);
+		drm_modeset_unlock(&drm->mode_config.connection_mutex);
 	}
 	switch (lcdc_type) {
 	case DRM_MODE_CONNECTOR_LVDS:
-- 
2.35.3

